lib/nerve.rb

```diff
@@ -86,15 +164,48 @@ def run
       $EXIT = true
     end
 
-    def launch_watcher(name, config)
-      log.debug "nerve: launching service watcher #{name}"
-      watcher = ServiceWatcher.new(config.merge({'instance_id' => @instance_id, 'name' => name}))
-      @watchers[name] = Thread.new{watcher.run}
+    def heartbeat
+      unless @heartbeat_path.nil?
+        FileUtils.touch(@heartbeat_path)
+      end
+      log.debug 'nerve: heartbeat'
+    end
+
+    def merged_config(config, name)
+      # Get a deep copy so sub-hashes are properly handled
+      deep_copy = Marshal.load(Marshal.dump(config))
+      return deep_copy.merge({'instance_id' => @instance_id, 'name' => name})
+    end
+
+    def launch_watcher(name, config, opts = {})
+      wait = opts[:wait] || false
+
+      watcher_config = merged_config(config, name)
+      # The ServiceWatcher may mutate the configs, so record the version before
+      # passing the config to the ServiceWatcher
+      @watcher_versions[name] = watcher_config.hash
+
+      watcher = ServiceWatcher.new(watcher_config)
+      unless @config_manager.options[:check_config]
+        log.debug "nerve: launching service watcher #{name}"
+        watcher.start()
+        @watchers[name] = watcher
+        if wait
+          log.info "nerve: waiting for watcher thread #{name} to report"
+          responsive_sleep(30) { !watcher.was_up.nil? || $EXIT }
+          log.info "nerve: watcher thread #{name} has reported!"
+        end
+      else
+        log.info "nerve: not launching #{name} due to --check-config option"
+      end
     end
 
     def reap_watcher(name)
-      watcher_thread = @watchers.delete(name)
-      watcher_thread.join()
+      watcher = @watchers.delete(name)
+      @watcher_versions.delete(name)
+      shutdown_status = watcher.stop()
+      log.info "nerve: stopped #{name}"
```

Comment:
Sure. Yea I don't actually know if/why they're happening, but we have seen quite a few nerve processes just spin in sched_wait which I suspect may be related (that's is what we added the heartbeat code to detect). While refactoring I noticed this join without a timeout and that's definitely a possibility.